src/lib.rs

```diff
@@ -1,21 +1,392 @@
+    /// tasks.
+    ///
+    /// Defaults to the number of logical CPUs on the machine.
+    ///
+    /// ### Panics
+    ///
+    /// Panics if `threads` is 0.
+    pub fn cpu_pool_threads(mut self, threads: usize) -> Self {
+        assert!(threads > 0);
+        self.cpu_pool_threads = threads;
+        self
+    }
+
+    /// Configure the size of mpsc buffers in the system.
+    ///
+    /// ### Panics
+    ///
+    /// Panics if `size` is 0.
+    pub fn channel_buffer_size(mut self, size: usize) -> Self {
+        assert!(size > 0);
+        self.channel_buffer_size = size;
+        self
+    }
+
+    /// Finish this `Options` builder and run the `starling` event loop with its
+    /// specified configuration.
+    pub fn run(self) -> Result<()> {
+        Starling::new(self)?.run()
+    }
+}
+
+impl Options {
+    // Get the number of `T`s that should be buffered in an mpsc channel for the
+    // current configuration.
+    fn buffer_capacity_for<T>(&self) -> usize {
+        let size_of_t = cmp::max(1, mem::size_of::<T>());
+        let capacity = self.channel_buffer_size / size_of_t;
+        cmp::max(1, capacity)
+    }
+}
+
+/// The Starling main thread.
+///
+/// The main thread doesn't do much other than supervise other threads: the IO
+/// event loop thread, various utility thread pools, and JavaScript task
+/// threads. Its primary responsibility is ensuring clean system shutdown and
+/// joining thread handles.
+pub(crate) struct Starling {
+    handle: StarlingHandle,
+    tasks: HashMap<task::TaskId, task::TaskHandle>,
+    threads: HashMap<thread::ThreadId, thread::JoinHandle<()>>,
+    receiver: Receiver<StarlingMessage>,
+}
+
+impl fmt::Debug for Starling {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Starling {{ .. }}")
+    }
+}
+
+impl Starling {
+    /// Construct a Starling system from the given options.
+    pub fn new(opts: Options) -> Result<Starling> {
+        let tasks = HashMap::new();
+        let threads = HashMap::new();
+
+        let (sender, receiver) = futures_mpsc::channel(opts.buffer_capacity_for::<StarlingMessage>());
+
+        let Options {
+            sync_io_pool_threads,
+            cpu_pool_threads,
+            ..
+        } = opts;
+
+        let handle = StarlingHandle {
+            options: Arc::new(opts),
+            sync_io_pool: CpuPool::new(sync_io_pool_threads),
+            cpu_pool: CpuPool::new(cpu_pool_threads),
+            sender,
+        };
+
+        Ok(Starling {
+            handle,
+            tasks,
+            threads,
+            receiver,
+        })
+    }
+
+    /// Run the main Starling event loop with the specified options.
+    pub fn run(mut self) -> Result<()> {
+        let (main, thread) =
+            task::Task::spawn_main(self.handle.clone(), self.handle.options().main.clone())?;
+
+        self.tasks.insert(main.id(), main.clone());
+        self.threads.insert(thread.thread().id(), thread);
```

Comment:
We're currently guaranteed to have a 1:1 mapping between tasks and threads but will eventually have an n:1 mapping, right? Worth adding a comment to that effect here, perhaps.

Reply:
Yep, this is just for the current setup, and yup that deserves a comment.